org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleBuild.java

```diff
@@ -55,4 +59,21 @@
      * @return the synchronization result
      */
     SynchronizationResult synchronize(IProgressMonitor monitor);
+
+    /**
+     * Executes an action in the Gradle runtime.
+     * <p>
+     * TODO (donat) explain how the action is executed (refer to Tooling API documentation)
+     * TODO (donat) explain common use-cases: task execution, test execution, and model loading
+     * TODO (donat) explain how the project connection is preconfigured: outputs and outputs to IDE console, progress and cancellation are wired to the monitor object
+     * TODO (donat) document exception handling
+     * TODO (donat) document scheduling rules
+     * TODO (donat) document how to load custom models (here or link to external documentation)
+     *
+     * @param action the action to execute
+     * @param monitor the monitor to report progress on, or {@code null} if progress reporting is not desired
+     * @return the result of the action
+     * @throws Exception when the action fails
+     */
+    <T> T withConnection(Function<ProjectConnection, ? extends T> action, IProgressMonitor monitor) throws Exception;
```

Comment:
This might be a personal preference, but I prefer "lambda parameters" as the last parameter.

```java
build.withConnection(connection -> {
    ...
}, monitor);
```

vs.

```java
build.withConnection(monitor, connection -> {
    ...
});
```

WDYT?

Reply:
Should we explicitly give the `action` access to the `monitor` by using a `BiFunction` or would we recommend just passing it the same `IProgressMonitor` instance?

```java
IProgressMonitor monitor = ...;
build.withConnection(connection -> {
    for (...) {
        monitor.worked(1);
    }
}, monitor);
```

Moreover, who's responsible for calling `monitor.done()`? The `action` or the code that executes the `action`?

Reply:
Re: argument order: the order is the same for most API methods in Eclipse. Here, I prefer consistency.

Reply:
Re: monitor object used by action: I think it's cleaner to use a simple `Function` and recommend the users to reuse `monitor` if needed within the action. Besides, we don't have to deal with calling `monitor.done()`. From the javadoc: _callers that pass a monitor are no longer allowed to rely upon the callee invoking done() and must do so themselves if necessary_.